Merge notebook into master for improved Notebook widget

RedFantom opened the pull request:
### PR Details:
- Widget name: `Notebook`
- Author: @j4321, with modifications by @Dogeek 

### Description
A `Notebook` widget implementation with functionality to reorder, close and detach tabs. In addition, when there are too many tabs, buttons appear to navigate through them. A `Menu` is also available to allow quick jumping to a specific tab.

### Checklist
- [x] Widget in a separate file in the appropriate folder
- [ ] Widget functions properly on both Windows and Linux
- [x] Widget code includes docstrings with parameter descriptions
- [x] Included an example file in `/examples`
- [ ] Widget is covered by unitttests in `/tests`
- [x] Widget includes required assets files
- [ ] Reference to widget in `AUTHORS.md`
- [ ] Entry in sphinx documentation

This PR contains much the same changes as #43 , but rebased onto `master`, changing commit messages and squashing commits into larger ones. This branch contains only changes required for the widget and no other commits.

RedFantom commented:
### TODO List:
- [ ] Extend the tests to properly cover the functionality of the widget
- [ ] Test the widget on Windows
- [x] Test the widget on Linux
- [ ] Create an entry in the sphinx documentation

RedFantom commented on tests/test_notebook.py:
```diff
@@ -0,0 +1,16 @@
+# -*- coding: utf-8 -*-
+"""
+Author : Dogeek
```
Please change the file docstring to either a single line or the standard format used in the other files in the repository.

RedFantom commented on tests/test_notebook.py:
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
```
The tests should be extended to be much more extensive and cover more of the `notebook.py` file. Just testing initialization is not sufficient.

RedFantom commented on ttkwidgets/notebook.py:
```diff
@@ -0,0 +1,935 @@
+#! /usr/bin/python3
```
Shebang lines are not necessary in code files like these.

RedFantom commented on ttkwidgets/notebook.py:
```diff
@@ -0,0 +1,935 @@
+#! /usr/bin/python3
+# -*- coding: utf-8 -*-
```
Encoding lines are only necessary in files that require them.

RedFantom commented on ttkwidgets/notebook.py:
```diff
@@ -0,0 +1,935 @@
+#! /usr/bin/python3
+# -*- coding: utf-8 -*-
+
```
Redundant whitespace.

RedFantom commented on ttkwidgets/notebook.py:
```diff
@@ -0,0 +1,935 @@
+        """
+        :param master: parent widget
+        :param tab_nb: tab index
+        :param **kwargs: keyword arguments for ttk::Frame widgets
```
These parameters are later passed to a `ttk.Label` initializer, note a `ttk.Frame` initializer. Is the docstring correct?

RedFantom commented on ttkwidgets/notebook.py:
```diff
@@ -0,0 +1,935 @@
+            self.closebtn.pack(side='right', padx=(0, 6))
+        self.update_idletasks()
+        self.configure(width=self.frame.winfo_reqwidth() + 6,
+                       height=self.frame.winfo_reqheight() + 6)
```
This line and the whole file may be wrapped to 120 characters.

RedFantom commented on ttkwidgets/notebook.py:
```diff
@@ -0,0 +1,935 @@
+        if self._state == 'disabled':
+            self.state(['disabled'])
+        elif self._state != 'normal':
+            raise ValueError("state option should be 'normal' or 'disabled'")
```
Raising of possible error should be mentioned in the docstring.

RedFantom commented on ttkwidgets/notebook.py:
```diff
@@ -0,0 +1,935 @@
+    _initialized = False
+
+    def __init__(self, master=None, **kwargs):
+        """
```
Docstring should be updated to support sphinx documentation.

RedFantom commented on ttkwidgets/notebook.py:
```diff
@@ -0,0 +1,935 @@
+                 'disabledfg': disabledfg,
+                 'disabledbg': disabledbg}
+
+        self.images = (
```
Instance attribute defined outside `__init__`. The existence of this attribute should be defined in `__init__`, though it may be initialized as an empty tuple.

RedFantom commented on ttkwidgets/notebook.py:
```diff
@@ -0,0 +1,935 @@
+
+    def insert(self, where, widget, **kwargs):
+        """
+        Insert WIDEGT at the position given by WHERE in the notebook.
```
Typo: `WIDEGT`

RedFantom commented on ttkwidgets/notebook.py:
```diff
@@ -0,0 +1,935 @@
+        nb.add(tab)
+        nb.grid()
+        self._toplevels.append(tl)
+        tl.mainloop()
```
`Toplevel.mainloop` need not be called: The `mainloop` of the master window is already running and if it is not, the user of the library is doing something wrong.

RedFantom commented on ttkwidgets/notebook.py:
```diff
@@ -0,0 +1,935 @@
+
+    def move_to_toplevel(self, tab):
+        tl = tk.Toplevel(self)
+        nb = Notebook(tl, **self._init_kwargs)
```
Should the tab still be contained in a `Notebook`? Dragging `Tab`s between two different `Notebook` widgets of two different parents is not entirely impossible yet still difficult to implement. Therefore, I propose that the `Frame` inside the `Notebook` is shown individually in its `Toplevel` without a `Notebook` until such a time that moving `Tab`s between different `Notebook`s is possible.

RedFantom commented on ttkwidgets/utilities.py:
```diff
@@ -9,3 +10,119 @@ def get_assets_directory():
+    return {key: widget.cget(key) for key in widget.keys()}
+
+
+def copy_widget(widget, new_parent, level=0):
```
This function and its dependency should be tested separately for all these test cases:
- [ ] Move a single widget from one parent to another
- [ ] Move a widget using grid on children from one parent to another
- [ ] Move a widget using pack on children from one parent to another
- [ ] Move a widget using place on children from one parent to another
- [ ] Move a widget using bindings from one parent to another
- [ ] Move a widget using a binding on its parent from one parent to another (and properly document its peculiarities)
- [ ] Attempt moving one widget to another `Tk` instance entirely (and throw an error if this cannot be done)

Dogeek replied:
AFAIK, `tkinter.Tk` is a singleton, so that last test doesn't matter

RedFantom replied:
Actually, there may be multiple `Tk` instances. It is a bad practice and makes running a program very complicated, but that is the point: It should be documented that this is not possible and the function should check whether that is actually what is requested and then throw a `RuntimeError` (or something). Then the test will determine whether the error is correctly thrown.

RedFantom commented on ttkwidgets/utilities.py:
```diff
@@ -9,3 +10,119 @@ def get_assets_directory():
+    return rv
+
+
+def parse_geometry(geometry):
```
Requires a separate unit test.

RedFantom commented on ttkwidgets/utilities.py:
```diff
@@ -9,3 +10,119 @@ def get_assets_directory():
+             int(match.group(1)), int(match.group(2)))
+
+
+def coordinates_in_box(coords, bbox, include_edges=True):
```
Requires a separate unit test. For example: Does it work with negative coordinates that may arise when using multiple monitors where the leftmost is not the primary monitor?

RedFantom commented:
Thank you for these commits, @Dogeek . I have cherry-picked them, which yields an easier and cleaner solution than merging into this branch. In the future, just mention the branch your commits are on, then I can continue cherry-picking them.

The CI is now failing because of issues with copying the widgets. If no value is given for an option, the default option (an empty value) appears to not always be valid. This will require some investigation.

Also, I would like to test the bindings in more detail, verifying that the actual functions are still bound. I will work on it when I can.

Dogeek commented:
@RedFantom I've fixed the copy_widget and move_widget functions in the `notebook-review-fixes` branch

RedFantom commented:
Thank you for taking a look at it! When I cherry-picked your commit and tested the solution against the unit tests, the issue still occurred however. I believe this is because the instantiation of an empty widget still leaves the options unspecified and rather than returning the defaults, `cget` will return an empty value.

These empty values must be filtered out. This does mean that if a widget is instantiated using an option that is an empty string, this will fail. However, I do not see any other option to do this properly.

Dogeek commented:
@RedFantom Added the fix with runtime error (same branch as before) and some more tests for utilities functions cause why not

Dogeek commented:
forgot to update the docstring for copy_widget (to signal possible raise of RuntimeError), docstring of move_widget could be updated as well

Dogeek commented:
I checked your commit for the get_widget_options thing, and my implementation is way better (dict comprehensions are about 10X faster than dict mutations, and I actually check if the option is at its default value instead of if the option is '' or None)

RedFantom commented:
> I checked your commit for the get_widget_options thing, and my implementation is way better (dict comprehensions are about 10X faster than dict mutations, and I actually check if the option is at its default value instead of if the option is '' or None)

While a comprehension may be faster, the solution you provide actually does not work, as the default values of the widget options are also empty. I checked by cherry-picking your commit and running the tests against it. My implementation might still be rewritten as a comprehension though.

Dogeek commented:
@RedFantom I've tested it myself, and it does work. it returns a dictionary of all the diverging keys from the default values. That means that if you do this :

```python
import tkinter as tk
root = tk.Tk()
lbl = tk.Label(root, takefocus=True)
print(get_widget_options(lbl))
# {'takefocus': 1}
```
you can see it works. If all the values are the default, it will return an empty dict, which is also fine.

Dogeek commented:
pushed a bunch of fixes to `notebook-review-fixes`

RedFantom commented:
My apologies, I must have cherry-picked and then merged the changes wrongly, as you are correct: Your implementation does indeed work. However, I decided to profile the function performance, and the version with dictionary modifications is significantly faster than the one with the dictionary comprehension. Even after taking out duplicate function calls and changing it so that the base widget object is only created once, the function with the 'normal' for loop yields a running time of `0ms` (the profiler doesn't measure smaller intervals) and for the dictionary comprehension I got it down to `2ms`.

I ran the test multiple times with multiple variations, but it was always slower. This is an interesting result that I had actually not expected. Perhaps the slow down is due to the added if-statement in the comprehension, I'm not sure.

Regardless, your implementation does work, but the one on the branch right now is a tad bit faster.

Dogeek commented:
Interesting. Though speed is not the only thing that matters in this case. I implemented it that way to take full advantage of Python's capabilities. My implementation is better in that regard, if the tkinter API ever changes, with the defaults changing, it's much better to do it my way.

Regardless, dict comprehensions are actually faster anyways, the only reason my implementation is slower is because there's 2 comprehensions, filtering 2 times. A single comprehension like :
```python
return {key: widget.cget(key) for key in widget.keys() if widget.cget(key)}
```
is surely faster than dict mutating.

RedFantom commented:
The tests added in commit [`a69f41f`](https://github.com/TkinterEP/ttkwidgets/pull/48/commits/a69f41fe029e08d5e8379d8d7ebd8be6dd69160d) illustrate my concerns about the `move_widget` function, and why I thought more testing was needed. Due to the way that commands and bound functions are referenced and executed in Tkinter, I thought that this would yield significant problems. It does indeed yield these problems. The only way I can see to circumvent the problem of function referencing in Tkinter is to somehow get the object instances of the bound functions and rebind them after the move is done. This is a critical thing: This way, when a widget is moved, then any bindings or commands on it are lost. This would mean that `Button`s also lose their functionality.

Of course, the widget can be made to not have the functionality of moving tabs to a new parent and in this way be merged.

The problem is illustrated by following code snippet:
```python
>>> import tkinter as tk
>>> window = tk.Tk()
>>> button = tk.Button(window, command=window.destroy
>>> button.cget("command")
'140439709217672destroy'
```

Dogeek commented:
The problem is that I don't know how to grab the reference to the functions.

The only solution I can see right now would be to hook into the ttk::Widget and tkinter::Widget classes to append the function references to a list. 

It should be that 'tk::Widget().bind(key)' returns the contents of the script but it doesn't.

@j4321 do you have any insight on this ?

RedFantom commented:
I thought I'd take a look at the Tkinter source code to figure out how the functions are called from the `Tcl`-interpreter. My research has so far lead to the following, for the Python 3.6 source tree:
- The `bind()` function is defined in the `Misc` class in [`__init__.py`](https://github.com/python/cpython/blob/3.6/Lib/tkinter/__init__.py#L1212).
- The `bind()` function calls the [`_bind()`](https://github.com/python/cpython/blob/3.6/Lib/tkinter/__init__.py#L1195) function on the same object.
- When the object given to bind to an event is actually a callable thing, it is registered using an ID with the function [`_register`](https://github.com/python/cpython/blob/3.6/Lib/tkinter/__init__.py#L1357)
- The `_register` function creates a new `Tcl` command (the text references that keep popping up in the errors) that calls the Python function using the [`CallWrapper`](https://github.com/python/cpython/blob/3.6/Lib/tkinter/__init__.py#L1692) class, which basically restores arguments and catches errors.
- The wrapped [`__call__`](https://github.com/python/cpython/blob/3.6/Lib/tkinter/__init__.py#L351) is passed to a function `_tk.createcommand(cbname, f)`, where `cbname` is the name that is returned by the `bind` function and the thing that keeps appearing.

This is where it gets... Complicated with the [`_tkinter.c`](https://github.com/python/cpython/blob/3.6/Modules/_tkinter.c#L2046) module. Let's continue.
- I believe the function called is [`_tkinter_tkapp_createcommand_impl`](https://github.com/python/cpython/blob/3.6/Modules/_tkinter.c#L2503) which matches the required signature. This function actually creates a function using the `Tcl_CreateCommand` function.
- The [`Tcl_CreateCommand`](https://core.tcl-lang.org/tcl/artifact/c2464e7067355a51) is a function included from `tcl.h` and is not part of the Python source code.
- The function passed to `Tcl_CreateCommand` is [`PythonCmd`](https://github.com/python/cpython/blob/3.6/Modules/_tkinter.c#L2400), together with a `PythonCmd_ClientData`, which contains the Python function, and other arguments.
- `PythonCmd` [unpacks](https://github.com/python/cpython/blob/3.6/Modules/_tkinter.c#L2426) the function and actually calls it with its arguments. Keep in mind that this is a `CallWrapper.__call__`, which unpacks the arguments and calls the function from Python.
- Critical is the [`destroy`](https://github.com/python/cpython/blob/3.6/Lib/tkinter/__init__.py#L584) function of the `Misc` function. When this function is called, it also deletes all commands that were part of the widget and for which the `needcleanup` kwarg of [`_register`](https://github.com/python/cpython/blob/3.6/Lib/tkinter/__init__.py#L1357) was set to `True`, or `1` by default.

### Conclusion
My conclusion is that once a binding is created, it can no longer be retrieved as a Python function, as it only still exists in the memory of the Python interpreter but there are no direct references to it from the Python. The Python interpreter simply keeps the memory as there is still a non-zero reference count to the created objects and the Tcl-interpreter manages its own memory, in which the binding is stored.

There are two ways I can see to change this behaviour:
- Override the `Misc.bind` function so that a separate reference is kept to the functions as well as the events they are bound to and by what arguments, so that they may be retrieved in a different format and rebound exactly as they were.
- Write a C-module extension that can take a Tcl-interpreter and so move the widget from there, also fetching commands and rebinding them. This would be a lot of work and pretty complex, if it is even possible at all to hook into Tkinter in such a way as none of its objects are declared in the header.

Note that I have not yet researched how the `command` kwarg for i.e. `Button`s is handled. I suspect these use the `_register` function directly, and moving them would require hooking into that function also.

Dogeek commented:
I found out that binding through that return value actually works for some reason. Like :

```python
import tkinter as tk


def callback(ev):
    print('Entered ' + str(ev.widget))


root = tk.Tk()
root.title('root')
frame = tk.Frame(root, width=300, height=300)
frame.bind('<Enter>', callback)
frame.pack()
print(frame.bind())  # ('<Enter>', )
print(frame.bind('<Button-1>'))  # 'if {"[64351688callback %# %b %f %h %k %s %t %w %x %y %A %E %K %N %W %T %X %Y %D]" == "break"} break\n'
tl = tk.Toplevel(root)
tl.title('tl')
other = tk.Frame(tl, width=300, height=300)
other.bind('<Enter>', frame.bind('<Enter>'))
other.pack()
root.mainloop()
```

The callback is still properly called, even though a string is passed as the callback argument. So I don't know why there is an error here

RedFantom commented:
Ah, of course. As you are invoking the `bind` function with string argument, that argument is directly regarded as a Tcl command and thus bound to that. I hadn't thought of that. That would fix the whole problem, as long as the bound function then has its reference count increased in the Tcl-interpreter and the Python-interpreter. This is something that has to be tested, but it could yield a viable solution as well.

So then the idea is to setup the bindings of the new widget to the string representations of the old widget's bindings. I'll try building an implementation.

RedFantom commented:
The solution for the bindings is actually simpler still. As the commands and their wrappers as far as I know do not specifically depend on the widgets that they are created for, they may be preserved by not destroying them when the widget is destroyed. The easiest way to do this is to make sure that the Python `Misc.destroy` function doesn't know about them. This way, they may still be invoked by the new widget and it does not create a memory leak as when the functions are unbound from the new widget or the new widget is destroyed, the wrappers are still destroyed.

Next up: Commands and the `place` geometry manager. I don't think I'll have any more time this weekend though.

Dogeek commented:
The place geometry manager is actually handled already.

The widgets are copied as is, with all their options. The command is actually behaving in the same way as the bind callbacks. I suppose every tkinter callback works that way (like after, bind, bind_all, command etc).

```py
>>> import tkinter as tk
>>> def cb():
...     print('callback')
...
>>> but = tk.Button(command=cb, text='ok')
>>> but.pack()
>>> callback

>>> but['command']
<string object: '44558152cb'>
>>> dir(but['command'])
['__add__', '__class__', '__contains__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getnewargs__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__mod__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__rmod__', '__rmul__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'capitalize', 'casefold', 'center', 'count', 'encode', 'endswith', 'expandtabs', 'find', 'format', 'format_map', 'index', 'isalnum', 'isalpha', 'isascii', 'isdecimal', 'isdigit', 'isidentifier', 'islower', 'isnumeric', 'isprintable', 'isspace', 'istitle', 'isupper', 'join', 'ljust', 'lower', 'lstrip', 'maketrans', 'partition', 'replace', 'rfind', 'rindex', 'rjust', 'rpartition', 'rsplit', 'rstrip', 'split', 'splitlines', 'startswith', 'strip', 'swapcase', 'title', 'translate', 'upper', 'zfill']
>>> but2 = tk.Button(command=but['command'], text='2')
>>> but2.pack()
>>> callback

>>> but2['command']
<string object: '44558152cb'>
>>>
```

RedFantom commented:
While it does appear to work the same way, the created Tcl function is not destroyed using the same function. If I run the following snippet, it still yields an error as the command callback has been destroyed.
```python
>>> import tkinter as tk
>>> def cb():
...     print("callback")
...
>>> but = tk.Button(command=cb, text="1")
>>> but2 = tk.Button(command=but.cget("command"), text="2")
>>> but.destroy()
>>> but2.invoke()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.6/tkinter/__init__.py", line 2388, in invoke
    return self.tk.call(self._w, 'invoke')
_tkinter.TclError: invalid command name "140345702173384cb"
```
This is also what the CI results show, on both Windows and Linux. So what must be identified is where the Tcl-command is destroyed for Buttons. This lies again in the Tkinter source code, and I will venture into it when I have the time to do so.

As for the `place`-geometry manager, an error was still occurring on the CI, so I thought I'd take a look at the tests. I extended them and discovered the issue with the `place` geometry manager. When the `place()` function is called with no arguments, it actually doesn't do anything. It doesn't place the widget in the parent window and it does not add it to the `place_slaves()` result. I added positional arguments, which fixed the problem.

I also implemented a `preserve_geometry` keyword argument, set to `False` by default, which makes the function automatically also position the new widget into its parent using the same geometry information as it was in the old parent with. 

Last but not least, I considered the possibility that a widget might not be configured using a geometry manager at all, and implemented a detection for that. It does require catching an error, which is slow, but the whole function is already slow anyway. The behaviour of `pack_info` is just not consistent with `grid_info` and `place_info`...

Dogeek commented:
Great job!

For the bind callback issue, I don't see a solution really, besides not destroying the widget, and using `pack/grid/place_forget` to just not show the widget which is not really a proper solution either because it would cause a memory leak if a widget is moved back and forth. Hooking into the bind function seems to be the only way to solve the issue. Then again, I hoped we'd find another solution, since hooks needs to be installed, and I hoped to not have to do that for obvious reasons (one being compatibility with other libraries, and the other being performance issues, even if the copy is slow in of itself, every millisecond matters)

I wholeheartedly agree with those changes. A couple of notes though ;

- You made the code python 2 compatible, and since we're dropping support for it, there's really no need

- You're importing TclError from tkinter, but the actual class in defined in the `_tkinter` module, which is the C extension module. It's closer to the source so it should be a tiny bit faster, with no downsides really.

Dogeek commented:
Also, in the docstrings, it's very minor, but the type is `tkinter.Widget` not `tk.Widget` It doesn't make much of a difference, but `tk` is an alias to `tkinter`, it would be better for the docs to refer to the actual module name, in my opinion.

RedFantom commented:
So, it turns out that the `command` kwarg value does indeed also get passed to the `_register` function through wrappers in the `Misc`-class. However, I had put the `_tclCommands = None` in a place so that it would only be executed when a binding is present. By moving this so that it is always executed, no commands are destroyed and thus now the `command` kwarg does work.

I'd rather keep referring as `tk.Widget`, as the tkinter module also uses that abbreviation for itself. Also, accessing a protected module such as `_tkinter` is bad practice, so I'd rather avoid and take the hit of accessing an additional pointer.

Now the only thing left to do here, I believe, is extend the tests of the `Notebook`-widget itself.

Dogeek commented:
What are the tests that you have in mind for the Notebook widget ? AFAIK there's not that much stuff to test for. If you provide a list, I can write those tests.

Also, do you have a better name for the `bbox_is_x1y1x2y2` keyword argument in the `coords_in_box()` function ? Couldn't find anything better when I wrote it but it's quite the mouthful.

RedFantom commented:
The name of the keyword argument `bbox_is_x1y1x2y2` is indeed a bit unfortunate, but even after giving it some thought, I haven't been able to think of a good replacement name. You could use `x1y2x2y2`, but that would be pretty unclear.

Most if not all of the functionality of the widget should be tested, insofar that is possible. So that would include:
- Adding tabs
- Removing tabs
- Moving/Dragging tabs
- The functions `insert` and `index`
- Adding too many tabs so the buttons appear
- The functionality of the buttons
- Detaching a tab
- Forgetting a tab
- Changing the options of an existing tab

Pretty much all the public functions at least. Some functionality such as dragging the tabs and such will be a bit complicated, so maybe not everything can be tested. The goal is to do as much as possible though.

Dogeek commented:
Okay, I'll start writing tests for these this weekend, or as soon as I figure out how to make my goddamn docs for my yandex python wrapper build on readthedocs.io. Whichever comes first.

RedFantom commented on ttkwidgets/notebook.py:
```diff
@@ -0,0 +1,926 @@
+        self.bind('<Control-Tab>', lambda e: self.select_next(True))
+        self.bind('<Shift-Control-ISO_Left_Tab>', lambda e: self.select_prev(True))
+
+    def index(self, tab_id):
```
While working on `test_notebook_insert`, I noticed a weird thing about this function. It does not return the index as it is visible to the user, rather is simply returns a valid `tab_id` given a host of different inputs. In fact, it does not even do this consistently: `tk.END` yields the length of `self._tabs`, which is not even a valid tab id or tab index!

I think the behaviour of this function must be consistent with that of `ttk.Notebook`, that is, it must return the index of the tab as it is visible to the user. To do this, it should use `self._visible_tabs`. As this function is used in a lot of other functions in the whole class, changing its behaviour completely would be ill-advised. Therefore, I think changing this function to be `_index()` and creating a new function `index()` is best. I have done this [here](https://github.com/TkinterEP/ttkwidgets/tree/notebook_index).

RedFantom commented:
The widget now looks like this, with the `black` theme applied:
![Screenshot_20200424_124707](https://user-images.githubusercontent.com/15170036/80204749-b8c63500-8629-11ea-8759-fb2d6c649022.png)
